core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/io/ImportRequest.java

```diff
@@ -0,0 +1,143 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.b2international.snowowl.core.request.io;
+
+import java.io.File;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import javax.validation.constraints.NotNull;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+
+import com.b2international.commons.CompareUtils;
+import com.b2international.commons.exceptions.ApiException;
+import com.b2international.snowowl.core.attachments.Attachment;
+import com.b2international.snowowl.core.attachments.AttachmentRegistry;
+import com.b2international.snowowl.core.attachments.InternalAttachmentRegistry;
+import com.b2international.snowowl.core.authorization.BranchAccessControl;
+import com.b2international.snowowl.core.domain.TransactionContext;
+import com.b2international.snowowl.core.identity.Permission;
+import com.b2international.snowowl.core.internal.locks.DatastoreLockContextDescriptions;
+import com.b2international.snowowl.core.request.LockRequest;
+import com.b2international.snowowl.core.uri.ComponentURI;
+import com.google.common.collect.Sets;
+
+/**
+ * @since 7.12
+ */
+public abstract class ImportRequest extends LockRequest<TransactionContext, ImportResponse> implements BranchAccessControl {
+	
+	private static final long serialVersionUID = 1L;
+	
+	@NotNull
+	private Attachment attachment;
+	
+	public ImportRequest() {
+		super(DatastoreLockContextDescriptions.IMPORT);
+	}
+	
+	public ImportRequest(String lockContext, String parentLockContext) {
+		super(lockContext, parentLockContext);
+	}
+	
+	final void setAttachment(Attachment attachment) {
+		this.attachment = attachment;
+	}
+	
+	protected Attachment attachment() {
+		return attachment;
+	}
+	
+	@Override
+	public final ImportResponse doExecute(TransactionContext context) {
+		context.log().info("Importing components from source file '%s'.", this.attachment.getFileName());
+		try {
+			InternalAttachmentRegistry iar = (InternalAttachmentRegistry) context.service(AttachmentRegistry.class);
+			File attachment = iar.getAttachment(this.attachment.getAttachmentId());
+			
+			ImportDefectAcceptor defectsAcceptor = new ImportDefectAcceptor(this.attachment.getFileName());
+			doValidate(context, attachment, defectsAcceptor, context.service(IProgressMonitor.class));
+			
+			final List<ImportDefect> defects = defectsAcceptor.getDefects();
+			if (!CompareUtils.isEmpty(defects)) {
+				return ImportResponse.defects(defects);
+			} else {
+				final Set<ComponentURI> visitedComponents = Sets.newHashSet();
+				doImport(context, attachment, visitedComponents::add, context.service(IProgressMonitor.class));
+				return ImportResponse.success(visitedComponents);
```

Comment:
...and then it would make sense to return them along with the components imported.